.. _rnaseqqcpipeline:

==============
3' Bias
==============

This report shows mean normalised read depth based on distance from
the 3' end of the transcript, in order to identify the 3' bias which
is common in polyA enriched RNA-seq.

Summary::
  * This analysis shows the normalised mean read depth for 3000 base
    pairs of each transcript starting from the 3'end.
  * The input is a database table generated by bam2profile.py.
  * Very high 3' bias may suggest a problem with the polyA enrichment
  * If there is 3' bias, it will artificially increase read depth for
    shorter transcripts.
  * More information is available here:
    http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-290

  These plots show the mean read depth of the 3' 3000 base pairs of
  the transcripts.  Transcripts longer than 3000 base pairs are
  cropped at this point, transcripts shorter than 3000 base pairs are excluded.


The good

.. report:: GoodExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

Ideally, these plots should show an even read depth along the read.


The bad

.. report:: BadExample.Tracker
   :render: myRenderer
   :transform: myTransform
   :options: myAesthetics

This plot shows data with a strong 3' bias.

More bad examples `<http://myBadData.html >`

Your data:

.. report:: RnaseqqcReport.ThreePrimeBias
   :render: r-ggplot
   :statement: aes(bin, counts) +
   	       geom_bar(size=0.3, stat="identity",
   	       fill="springgreen4", colour="springgreen4") +
	       scale_color_brewer(palette = "Set1") +
	       theme(panel.background=element_rect(fill="white",
			colour="white"), legend.position="none") +
	       ggtitle("3 Prime Bias") +
	       scale_x_discrete(name="Distance from 3 prime end",
	       			labels=c(-3000, -2500, -2000, -1500,
   						-1000, -500, "3\nprime\nend"),
   	       breaks=c(0, 500, 1000, 1500, 2000, 2500, 3000)) +
	       ylab("Normalised Read Count")


Commentary
  This will take the form of some active comments.  This will require the report to
  be published so that it is hosted on the CGAT server/ comments on the DISQUS server.

